cql3/expr/expression.cc

```diff
@@ -1911,6 +1914,71 @@ utils::chunked_vector<std::vector<managed_bytes_opt>> get_list_of_tuples_element
     return tuples_list;
 }
 
+void fill_prepare_context(expression& e, prepare_context& ctx) {
+    expr::visit(overloaded_functor {
+        [&](bind_variable& bind_var) {
+            ctx.add_variable_specification(bind_var.bind_index, bind_var.receiver);
+        },
+        [&](collection_constructor& c) {
+            for (expr::expression& element : c.elements) {
+                fill_prepare_context(element, ctx);
+            }
+        },
+        [&](tuple_constructor& t) {
+            for (expr::expression& element : t.elements) {
+                fill_prepare_context(element, ctx);
+            }
+        },
+        [&](usertype_constructor& u) {
+            for (auto& [field_name, field_val] : u.elements) {
+                fill_prepare_context(field_val, ctx);
+            }
+        },
+        [&](function_call& f) {
+            const shared_ptr<functions::function>& func = std::get<shared_ptr<functions::function>>(f.func);
+            if (ctx.is_processing_pk_restrictions() && !func->is_pure()) {
+                ctx.add_pk_function_call(f);
+            }
```

Comment:
I'm convincing myself (it's not too hard) that we need a separate pass to prepare just the partition key functions. We're not interested in functions in general, just non-pure functions in the partition key expression of a modification_statement that is an LWT. Seems simpler to move the responsibility there.

Reply:
We can even think of splitting the expression into two:

```c++
struct bind_variable_assignment {
    unsigned bind_variable_id;
    expression expr;
}
```

class modification_statement {
    std::vector<bind_variable_assignment> _pre_bounce_bind_variable_assignments;
    ...
};
```

Then we convert non-pure function calls to a bind_variable_assignment, and replace the call with a reference to the new bind variables. I don't propose to do this, just an exercise in thinking how it could evolve.